sys/include/frac.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2018 Eistec AB
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup  sys_frac   Fractional integer operations
+ * @ingroup   sys
+ *
+ * This header provides some functions for scaling integers by fractions, while
+ * preserving as many bits as possible.
+ *
+ * The implementation requires that @ref frac_t is initialized properly, either
+ * by calling @ref frac_init, which will compute the algorithm parameters at
+ * runtime, or via a precomputed initializer.
+ *
+ * Precomputing the frac_t values can be done via the application found in
+ * `tests/frac-config` in the RIOT tree.
+ *
+ * ### Numeric precision
+ *
+ * The algorithm will under certain circumstances give an incorrectly rounded
+ * result, more precisely, the result may sometimes be rounded up instead of
+ * rounded down when the product in the numerator, @$p = x \cdot num@$, would
+ * result in @$p >= 2^{31}@$. Fortunately, the relative error of this rounding
+ * mistake is small.
+ *
+ * This tradeoff is a design choice to make the algorithm faster.
+ *
+ * @see       Libdivide homepage: http://libdivide.com/
+ *
+ * @file
+ * @ingroup   sys
+ * @author    Joakim Nohlgård <joakim.nohlgard@eistec.se>
+ * @{
+ */
+
+#ifndef FRAC_H
+#define FRAC_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   frac descriptor for fraction consisting of two 32 bit integers
+ */
+typedef struct {
+    uint32_t frac; /**< fraction */
+    uint8_t shift; /**< exponent */
+} frac_t;
+
+/**
+ * @brief   Initialize frac_t struct
+ *
+ * This function computes the mathematical parameters used by the frac algorithm.
+ *
+ * @note Be extra careful if @p num > @p den, the result from @ref frac_scale
```

Comment:
@gebart if you're still reading github notifications :wink: - doesn't the result end up being `(x * den / nom) % 32`, as if calculated using 32bit floats?

Reply:
it does.